python/ray/tune/sync_client.py

```diff
@@ -418,3 +424,228 @@ def _validate_exclude_template(exclude_template):
+                f"Last remote task sync still in progress, "
+                f"skipping sync from {source} to {target}."
+            )
+            return False
+
+        source_ip, source_path = source
+        target_ip = ray.util.get_node_ip_address()
+
+        self._last_source_tuple = source_ip, source_path
+        self._last_target_tuple = target_ip, target
+
+        # Get existing files on local node before packing on remote node
+        self._last_files_stats = _get_recursive_files_and_stats(target)
+
+        return self._execute_sync(
+            self._last_source_tuple, self._last_target_tuple, self._last_files_stats
+        )
+
+    def sync_up(
+        self, source: str, target: Tuple[str, str], exclude: Optional[List] = None
+    ) -> bool:
+        if self._sync_still_running():
+            logger.warning(
+                f"Last remote task sync still in progress, "
+                f"skipping sync from {source} to {target}."
+            )
+            return False
+
+        source_ip = ray.util.get_node_ip_address()
+        target_ip, target_path = target
+
+        self._last_source_tuple = source_ip, source
+        self._last_target_tuple = target_ip, target_path
+
+        # Get existing files on remote node before packing on local node
+        self._last_files_stats = _remote_get_recursive_files_and_stats.options(
+            num_cpus=0, resources={f"node:{target_ip}": 0.01}
+        ).remote(target_path)
+
+        return self._execute_sync(
+            self._last_source_tuple, self._last_target_tuple, self._last_files_stats
+        )
+
+    def _execute_sync(
+        self,
+        source_tuple: Tuple[str, str],
+        target_tuple: Tuple[str, str],
+        files_stats: Optional[Dict[str, Tuple[float, int]]] = None,
+    ) -> bool:
+        source_ip, source_path = source_tuple
+        target_ip, target_path = target_tuple
+
+        pack_on_source_node = _pack_dir.options(
+            num_cpus=0, resources={f"node:{source_ip}": 0.01}
+        )
+        unpack_on_target_node = _unpack_dir.options(
+            num_cpus=0, resources={f"node:{target_ip}": 0.01}
+        )
+
+        pack_future = pack_on_source_node.remote(source_path, files_stats)
+        if self._store_pack_future:
+            self._pack_future = pack_future
+        self._sync_future = unpack_on_target_node.remote(pack_future, target_path)
+        return True
+
+    def delete(self, target: str):
+        pass
+
+    def wait(self):
+        if self._sync_future:
+            try:
+                ray.get(self._sync_future)
+            except Exception as e:
+                raise TuneError(
+                    f"Remote task sync failed from "
+                    f"{self._last_source_tuple} to "
+                    f"{self._last_target_tuple}: {e}"
+                ) from e
+            self._sync_future = None
+            self._pack_future = None
+
+    def wait_or_retry(self, max_retries: int = 3, backoff_s: int = 5):
```

Comment:
wait_and_retry()? it seems we do both here.

Reply:
Ah it's more like call `wait()` or if it fails, retry (we don't retry if wait succeeds, so not always "and"). Also just implementing the higher level API (and the code is mostly copied from `CommandBasedClient`(